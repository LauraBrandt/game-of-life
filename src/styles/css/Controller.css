.controller .controller-controls, .controller .controller-start-config {
  background-color: #ffc30f;
  border-radius: 8px;
  margin: 1rem;
  padding: 0.5em 0; }
  .controller .controller-controls p, .controller .controller-start-config p {
    padding: 0 0.5em 0.4em 0.5em;
    margin: 0; }

.controller {
  font-family: "Montserrat", sans-serif;
  display: flex;
  margin: 0 1em; }
  @media screen and (max-width: 770px) {
    .controller {
      font-size: 0.9em;
      flex-direction: column; } }
  .controller .controller-controls {
    flex: 4;
    height: max-content; }
    .controller .controller-controls .arrows {
      display: flex;
      justify-content: space-around;
      align-items: center; }
      .controller .controller-controls .arrows p {
        margin: 0;
        padding: 0;
        flex: 3;
        text-align: center; }
      .controller .controller-controls .arrows .arrow {
        font-size: 1.3rem;
        color: white;
        padding: 0.2rem;
        width: 2.5rem;
        height: 2.5rem;
        margin: 0.3rem;
        font-family: "Montserrat", sans-serif;
        border: 1px solid #aaa;
        border-radius: 5px;
        background-color: #e5ac00;
        cursor: pointer;
        flex: 1;
        align-self: stretch; }
        .controller .controller-controls .arrows .arrow:hover, .controller .controller-controls .arrows .arrow:focus {
          outline: none;
          background-color: #efb400; }
        .controller .controller-controls .arrows .arrow:active {
          font-size: 1.2rem;
          background-color: #fabb00; }
        @media screen and (max-width: 770px) {
          .controller .controller-controls .arrows .arrow {
            font-size: 0.9em; } }
  .controller .controller-start-config {
    flex: 5; }
    .controller .controller-start-config span {
      color: rgba(0, 0, 0, 0.7); }
  .controller .controller-button {
    display: block;
    width: 100%;
    color: white;
    border: none;
    padding: 0.4em;
    height: 2.1rem;
    font-size: 1em;
    font-family: "Montserrat", sans-serif;
    border-bottom: 1px solid #aaa;
    background-color: #e5ac00;
    cursor: pointer; }
    .controller .controller-button:nth-child(2) {
      border-top: 1px solid #aaa; }
    .controller .controller-button:hover, .controller .controller-button:focus {
      outline: none;
      background-color: #efb400; }
    .controller .controller-button:active {
      font-size: 0.95em;
      background-color: #fabb00; }
    @media screen and (max-width: 770px) {
      .controller .controller-button {
        font-size: 0.9em; } }
  .controller .controller-pattern-dropdown {
    list-style-type: none;
    margin: 0;
    padding: 0; }
    .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button {
      display: block;
      width: 90%;
      margin: auto;
      color: #fafafa;
      border: none;
      padding: 0.3em;
      height: 1.8rem;
      font-size: 0.9em;
      font-family: "Montserrat", sans-serif;
      background-color: #efb400;
      border-bottom: 1px solid #aaa;
      cursor: pointer; }
      .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:hover, .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:focus {
        outline: none;
        background-color: #fabb00; }
      .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:active {
        font-size: 0.85em;
        background-color: #ffc005; }
      @media screen and (max-width: 770px) {
        .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button {
          font-size: 0.8em; } }
