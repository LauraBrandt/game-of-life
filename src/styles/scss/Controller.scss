@import 'variables';

%controller-div {
  background-color: $yellow;
  border-radius: 8px;
  margin: 1rem;
  padding: 0.5em 0;
  p {
    padding: 0 0.5em 0.4em 0.5em;
    margin: 0;
  }
}

.controller {
  font-family: $main-font;
  display: flex;
  margin: 0 1em;
  @media screen and (max-width: $break-sm) {
    font-size: 0.9em;
    flex-direction: column;
  }

  .controller-controls {
    @extend %controller-div;
    flex: 4;
    height: max-content; 
  }
  .controller-start-config {
    @extend %controller-div;
    flex: 5;
    span {
      color: rgba(0,0,0,0.7);
    }
  }
  .controller-button {
    display: block;
    width: 100%;
    color: white;
    border: none;
    padding: 0.4em;
    height: 2.1rem;
    font-size: 1em;
    font-family: $main-font;
    border-bottom: 1px solid $grey;
    background-color: darken($yellow, 8%);
    cursor: pointer;
    &:nth-child(2) {
      border-top: 1px solid $grey;
    }
    &:hover, &:focus {
      outline: none;
      background-color: darken($yellow, 6%);
    }
    &:active {
      font-size: 0.95em;
      background-color: darken($yellow, 4%);
    }
    @media screen and (max-width: $break-sm) {
      font-size: 0.9em;
    }
  }
  .controller-pattern-dropdown {
    list-style-type: none;
    margin: 0;
    padding: 0;
    .controller-dropdown-element {
      .controller-dropdown-button {
        display: block;
        width: 90%;
        margin: auto;
        color: $light;
        border: none;
        padding: 0.3em;
        height: 1.8rem;
        font-size: 0.9em;
        font-family: $main-font;
        background-color: darken($yellow, 6%);
        border-bottom: 1px solid $grey;
        cursor: pointer;
        &:hover, &:focus {
          outline: none;
          background-color: darken($yellow, 4%);
        }
        &:active {
          font-size: 0.85em;
          background-color: darken($yellow, 2%);
        }
        @media screen and (max-width: $break-sm) {
          font-size: 0.8em;
        }
      }
    }
  }
}