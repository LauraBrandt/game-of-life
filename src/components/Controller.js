import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { tick, back, setRandom, setPattern, clearBoard } from '../actions/boardActions';
import { incrementGeneration, decrementGeneration, resetGeneration } from '../actions/generationActions';
import '../styles/css/Controller.css';
import patterns from '../startPatterns';

export class Controller extends Component {
  constructor() {
    super();

    this.state = {
      timer: null,
      isPlaying: false,
      dropdownOpen: false,
      patterns: []
    }

    this.togglePlay = this.togglePlay.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.clear = this.clear.bind(this);
    this.randomize = this.randomize.bind(this);
    this.setPattern = this.setPattern.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
    this.goBack = this.goBack.bind(this);
    this.goForward = this.goForward.bind(this);
  }

  togglePlay() {
    if (this.state.isPlaying) {
      clearInterval(this.state.timer);
    } else {
      this.startTimer();
    }
    this.setState( prevState => ({ isPlaying: !prevState.isPlaying }) );
  }

  startTimer() {
    const timer = setInterval(() => {
      this.props.tick();
      this.props.incrementGeneration();
    }, 250);
    this.setState({ timer });
  }

  clear() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.clearBoard();
    this.props.resetGeneration();
  }

  randomize() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.randomize();
  }

  setPattern(cells) {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.setPattern(cells);
  }

  toggleDropdown() {
    this.setState(prevState => ({dropdownOpen: !prevState.dropdownOpen}));
  }

  componentDidMount() {
    this.setState({ patterns });
    this.randomize();
  }

  componentWillUnmount() {
    clearInterval(this.state.timer);
  }

  goBack() {
    this.props.back();
    this.props.decrementGeneration();
  }

  goForward() {
    this.props.tick();
    this.props.incrementGeneration();
  }

  render() {
    return (
      <div className="controller">
        <div className="controller-controls">
          <p>Controls</p>
          <button className="controller-button" onClick={this.togglePlay}>
            {this.state.isPlaying ? 'Pause' : 'Play'}
          </button>
          <button className="controller-button" onClick={this.clear}>
            Clear
          </button>
          <div className="arrows">
            <button className="arrow arrow-left" onClick={this.goBack}>&#9664;</button>
            <p>Move 1 generation</p>
            <button className="arrow arrow-right" onClick={this.goForward}>&#9654;</button>
          </div>
        </div>
        
        <div className="controller-start-config">
          <p>Choose a starting configuration: <br/><span>(or click the squares on the board to make your own)</span></p>
          <button className="controller-button" onClick={this.randomize}>
            Random
          </button>
          <div className="controller-patterns">
            <button className="controller-button controller-pattern" onClick={this.toggleDropdown}>
              Patterns &#9662;
            </button>
            {this.state.dropdownOpen &&
              <ul className="controller-pattern-dropdown"> 
                {this.state.patterns.map(pattern => 
                  <li key={pattern.code} className="controller-dropdown-element">
                    <button className="controller-dropdown-button" onClick={this.setPattern.bind(null, pattern.cells)}>
                      {pattern.name}
                    </button>
                  </li>
                )}
              </ul>
            }
          </div>
        </div>
      </div>
    );
  }
}

Controller.propTypes = {
  tick: PropTypes.func.isRequired,
  back: PropTypes.func.isRequired,
  randomize: PropTypes.func.isRequired,
  setPattern: PropTypes.func.isRequired,
  clearBoard: PropTypes.func.isRequired,
  incrementGeneration: PropTypes.func.isRequired,
  decrementGeneration: PropTypes.func.isRequired,
  resetGeneration: PropTypes.func.isRequired
}

const mapDispatchToProps = dispatch => ({
  randomize: () => dispatch(setRandom()),
  setPattern: pattern => dispatch(setPattern(pattern)),
  tick: () => dispatch(tick()),
  back: () => dispatch(back()),
  clearBoard: () => dispatch(clearBoard()),
  incrementGeneration: () => dispatch(incrementGeneration()),
  decrementGeneration: () => dispatch(decrementGeneration()),
  resetGeneration: () => dispatch(resetGeneration()),
});

export default connect(null, mapDispatchToProps)(Controller);
