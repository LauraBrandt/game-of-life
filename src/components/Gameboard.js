import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toggleCell } from '../actions/boardActions';
import '../styles/css/Gameboard.css';

export class Gameboard extends Component {
  render() {
    const { board, toggleCell } = this.props;

    return (
      <div className="gameboard">
        { board.map((row, i) => 
          <div className="board-row" key={i}>
            {row.map((cell, j) => 
              <div 
                className={`board-cell${cell ? ' active' : ''}`} 
                key={j} 
                onClick={toggleCell.bind(null, [i, j])}>
              </div>
            )}
          </div>
        )}
      </div>
    );
  }
}

Gameboard.propTypes = {
  board: PropTypes.array.isRequired,
  toggleCell: PropTypes.func
}

const mapStateToProps = state => ({
  board: state.board.current
});

const mapDispatchToProps = dispatch => ({
  toggleCell: cell => dispatch(toggleCell(cell))
});

export default connect(mapStateToProps, mapDispatchToProps)(Gameboard);
