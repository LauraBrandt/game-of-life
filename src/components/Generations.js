import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import '../styles/css/Generations.css';

export class Generations extends Component {
  render() {
    return (
      <div className="generations">
        Generation: <span>{this.props.generation}</span>
      </div>
    );
  }
}

Generations.propTypes = {
  generation: PropTypes.number.isRequired
}

const mapStateToProps = state => ({
  generation: state.generation
});

export default connect(mapStateToProps)(Generations);
