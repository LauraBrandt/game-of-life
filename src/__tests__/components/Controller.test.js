import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import { Controller } from '../../components/Controller';

describe('<Controller/>', () => {
  const props = {
    tick: jest.fn(),
    randomize: jest.fn(),
    setPattern: jest.fn(),
    clearBoard: jest.fn(),
    incrementGeneration: jest.fn(),
    resetGeneration: jest.fn()
  }

  let wrapper;

  afterEach(() => {
    props.tick.mockClear();
    props.randomize.mockClear();
    props.setPattern.mockClear();
    props.clearBoard.mockClear();
    props.incrementGeneration.mockClear();
    props.resetGeneration.mockClear();
  });

  describe('content', () => {
    beforeEach(() => {
      wrapper = shallow(<Controller {...props}/>);
    });

    it('has 2 child divs', () => {
      expect(wrapper.children()).toHaveLength(2);
    });

    it('has the correct number of buttons', () => {
      expect(wrapper.find('.controller-button')).toHaveLength(4);
    });

    it('the first button has text "Play" or "Pause" depending on state.isPlaying', () => {
      wrapper.setState({ isPlaying: false });
      let button = wrapper.find('.controller-button').at(0);
      expect(button.text()).toBe('Play');
      
      wrapper.setState({ isPlaying: true });
      button = wrapper.find('.controller-button').at(0);
      expect(button.text()).toBe('Pause');
    });

    it('the second button has text "Clear"', () => {
      const button = wrapper.find('.controller-button').at(1);
      expect(button.text()).toBe('Clear');
    });

    it('the third button has text "Random"', () => {
      const button = wrapper.find('.controller-button').at(2);
      expect(button.text()).toBe('Random');
    });

    it('the fourth button has text "Patterns"', () => {
      const button = wrapper.find('.controller-button').at(3);
      expect(button.text()).toContain('Patterns');
    });

    it("the dropdown doesn't show unless state.dropdownOpen = true", () => {
      wrapper.setState({ dropdownOpen: false });
      expect(wrapper.find('ul')).toHaveLength(0);
      
      wrapper.setState({ dropdownOpen: true });
      expect(wrapper.find('ul')).toHaveLength(1);
    });

    it('the dropdown has the correct number of list items', () => {
      wrapper.setState({ dropdownOpen: true });
      const numPatterns = wrapper.state('patterns').length;
      const numLi = wrapper.find('ul').find('li').length;
      expect(numLi).toBe(numPatterns);
    });
  });

  describe('interactions', () => {
    beforeEach(() => {
      wrapper = shallow(<Controller {...props}/>);
    });

    it('calls togglePlay on click of Play button', () => {
      wrapper.instance().togglePlay = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.update();

      wrapper.find('.controller-button').at(0).simulate('click');
      
      expect(wrapper.instance().togglePlay).toBeCalled();
    });

    it('calls clear on click of Clear button', () => {
      wrapper.instance().clear = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.update();

      wrapper.find('.controller-button').at(1).simulate('click');
      
      expect(wrapper.instance().clear).toBeCalled();
    });

    it('calls randomize on click of Random button', () => {
      wrapper.instance().randomize = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.update();

      wrapper.find('.controller-button').at(2).simulate('click');
      
      expect(wrapper.instance().randomize).toBeCalled();
    });

    it('calls toggleDropdown on click of Patterns button', () => {
      wrapper.instance().toggleDropdown = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.update();

      wrapper.find('.controller-pattern').simulate('click');
      
      expect(wrapper.instance().toggleDropdown).toBeCalled();
    });

    it('calls setPattern with correct argument on click of one of the pattern dropdown buttons', () => {
      wrapper.instance().setPattern = jest.fn();
      wrapper.setState({ 
        patterns: [
          {
            name: "Test Pattern",
            code: "testpattern",
            cells: [
              [5, 1], [3, 2]
            ]
          },
        ],
        dropdownOpen: true
      });

      const button = wrapper.find('.controller-pattern-dropdown').find('button').at(0);
      button.simulate('click');
      
      expect(wrapper.instance().setPattern).toBeCalled();
      expect(wrapper.instance().setPattern).toBeCalledWith([[5, 1], [3, 2]]);
    });
  });

  describe('methods', () => {
    beforeEach(() => {
      jest.useFakeTimers();
      wrapper = shallow(<Controller {...props}/>);
    });

    afterEach(() => {
      jest.clearAllTimers();
    });

    describe('componentDidMount', () => {
      it('sets the patterns when mounted', () => {
        expect(wrapper.state('patterns').length).toBeGreaterThan(0);
      });

      it('calls randomize when mounted', () => {
        spyOn(Controller.prototype, 'randomize')
        mount(<Controller {...props}/>);

        expect(Controller.prototype.randomize).toHaveBeenCalledTimes(1);
      });
    });

    describe('startTimer', () => {
      it('calls setInterval', () => {
        wrapper.instance().startTimer();
        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 250);
      });
      
      it('calls props.tick and props.incrementGeneration', () => {
        wrapper.instance().startTimer();

        expect(wrapper.instance().props.tick).not.toBeCalled();
        expect(wrapper.instance().props.incrementGeneration).not.toBeCalled();

        jest.runOnlyPendingTimers();

        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); 
        expect(wrapper.instance().props.incrementGeneration).toHaveBeenCalledTimes(1); 
      });

      it('calls runs every 250ms', () => {
        wrapper.instance().startTimer();
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(0);
        expect(wrapper.instance().props.incrementGeneration).toHaveBeenCalledTimes(0);
        
        jest.runTimersToTime(300);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().props.incrementGeneration).toHaveBeenCalledTimes(1);
        
        jest.runTimersToTime(250);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(2); 
        expect(wrapper.instance().props.incrementGeneration).toHaveBeenCalledTimes(2); 
        
        jest.runTimersToTime(2000);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(10); 
        expect(wrapper.instance().props.incrementGeneration).toHaveBeenCalledTimes(10); 
      });

      it('sets state.timer', () => {
        expect(wrapper.state('timer')).toBeNull();
        wrapper.instance().startTimer();
        expect(wrapper.state('timer')).not.toBeNull();
      });
    });

    describe('togglePlay', () => {
      it('clears timer if state.isPlaying is true', () => {
        wrapper.setState({ isPlaying: true });
        wrapper.instance().startTimer();

        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(0); // beginning of timer
        jest.runTimersToTime(300);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // timer runs once
        
        wrapper.instance().togglePlay();
        jest.runTimersToTime(250);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // not called again       
      });
      
      it('starts timer if state.isPlaying is false', () => {
        wrapper.instance().startTimer = jest.fn();
        wrapper.setState({ isPlaying: false });

        expect(wrapper.instance().startTimer).not.toBeCalled();

        wrapper.instance().togglePlay();

        expect(wrapper.instance().startTimer).toBeCalled();
      });

      it('toggles state.isPlaying', () => {
        wrapper.setState({ isPlaying: false });

        wrapper.instance().togglePlay();
        expect(wrapper.state('isPlaying')).toBe(true);
        
        wrapper.instance().togglePlay();
        expect(wrapper.state('isPlaying')).toBe(false);
      });
    });

    describe('clear', () => {
      it('clears the timer', () => {
        wrapper.instance().startTimer();
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(0); // beginning of timer
        jest.runTimersToTime(300);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // timer runs once
        
        wrapper.instance().clear();
        jest.runTimersToTime(250);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // not called again
      });

      it('changes state.isPlaying to false', () => {
        wrapper.setState({ isPlaying: true });
        wrapper.instance().clear();
        expect(wrapper.state('isPlaying')).toBe(false);
      });

      it('calls props.clearBoard', () => {
        const initial = wrapper.instance().props.clearBoard.mock.calls.length;
        wrapper.instance().clear();
        expect(wrapper.instance().props.clearBoard).toHaveBeenCalledTimes(initial + 1);
      });

      it('calls props.resetGeneration', () => {
        const initial = wrapper.instance().props.resetGeneration.mock.calls.length;
        wrapper.instance().clear();
        expect(wrapper.instance().props.resetGeneration).toHaveBeenCalledTimes(initial + 1);
      });
    });

    describe('randomize', () => {
      it('clears the timer', () => {
        wrapper.instance().startTimer();
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(0); // beginning of timer
        jest.runTimersToTime(300);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // timer runs once
        
        wrapper.instance().randomize();
        jest.runTimersToTime(250);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // not called again
      });

      it('changes state.isPlaying to false', () => {
        wrapper.setState({ isPlaying: true });
        wrapper.instance().randomize();
        expect(wrapper.state('isPlaying')).toBe(false);
      });

      it('calls props.resetGeneration', () => {
        const initial = wrapper.instance().props.resetGeneration.mock.calls.length;
        wrapper.instance().randomize();
        expect(wrapper.instance().props.resetGeneration).toHaveBeenCalledTimes(initial + 1);
      });

      it('calls props.randomize', () => {
        const initial = wrapper.instance().props.randomize.mock.calls.length;
        wrapper.instance().randomize();
        expect(wrapper.instance().props.randomize).toHaveBeenCalledTimes(initial + 1);
      });
    });

    describe('setPattern', () => {
      const pattern = [
        [1,2], [3,4]
      ];

      it('clears the timer', () => {
        wrapper.instance().startTimer();
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(0); // beginning of timer
        jest.runTimersToTime(300);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // timer runs once
        
        wrapper.instance().setPattern(pattern);
        jest.runTimersToTime(250);
        expect(wrapper.instance().props.tick).toHaveBeenCalledTimes(1); // not called again
      });

      it('changes state.isPlaying to false', () => {
        wrapper.setState({ isPlaying: true });
        wrapper.instance().setPattern(pattern);
        expect(wrapper.state('isPlaying')).toBe(false);
      });

      it('calls props.resetGeneration', () => {
        const initial = wrapper.instance().props.resetGeneration.mock.calls.length;
        wrapper.instance().setPattern(pattern);
        expect(wrapper.instance().props.resetGeneration).toHaveBeenCalledTimes(initial + 1);
      });

      it('calls props.setPattern', () => {
        const initial = wrapper.instance().props.setPattern.mock.calls.length;
        wrapper.instance().setPattern(pattern);
        expect(wrapper.instance().props.setPattern).toHaveBeenCalledTimes(initial + 1);
        expect(wrapper.instance().props.setPattern).toBeCalledWith([[1,2],[3,4]]);
      });
    });

    describe('toggleDropdown', () => {
      it('toggles state.dropdownOpen', () => {
        expect(wrapper.state('dropdownOpen')).toBe(false);
        wrapper.instance().toggleDropdown();
        expect(wrapper.state('dropdownOpen')).toBe(true);
      });
    });

    describe('componentWillUnmount', () => {
      it('clears timer on unmount', () => {
        wrapper.instance().startTimer();
        expect(props.tick).toHaveBeenCalledTimes(0); // beginning of timer
        jest.runTimersToTime(300);
        expect(props.tick).toHaveBeenCalledTimes(1); // timer runs once
        
        wrapper.unmount();

        jest.runTimersToTime(250);
        expect(props.tick).toHaveBeenCalledTimes(1); // not called again
      });
    });
  });

  describe('snapshot', () => {
    it('renders correctly', () => {
      const component = renderer.create(<Controller {...props}/>);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});