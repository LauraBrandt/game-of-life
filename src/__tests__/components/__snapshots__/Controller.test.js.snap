// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controller/> snapshot renders correctly 1`] = `
<div
  className="controller"
>
  <div
    className="controller-controls"
  >
    <p>
      Controls
    </p>
    <button
      className="controller-button"
      onClick={[Function]}
    >
      Play
    </button>
    <button
      className="controller-button"
      onClick={[Function]}
    >
      Clear
    </button>
    <div
      className="arrows"
    >
      <button
        className="arrow arrow-left"
        onClick={[Function]}
      >
        ◀
      </button>
      <p>
        Move 1 generation
      </p>
      <button
        className="arrow arrow-right"
        onClick={[Function]}
      >
        ▶
      </button>
    </div>
  </div>
  <div
    className="controller-start-config"
  >
    <p>
      Choose a starting configuration: 
      <br />
      <span>
        (or click the squares on the board to make your own)
      </span>
    </p>
    <button
      className="controller-button"
      onClick={[Function]}
    >
      Random
    </button>
    <div
      className="controller-patterns"
    >
      <button
        className="controller-button controller-pattern"
        onClick={[Function]}
      >
        Patterns ▾
      </button>
    </div>
  </div>
</div>
`;
